ovn_k8s/modes/overlay.py

```diff
@@ -304,3 +312,124 @@ def add_endpoint(self, event):
+        tcp_nodeport_services = []
+        udp_nodeport_services = []
+        tcp_services = []
+        udp_services = []
+        if services and 'items' in services:
+            for service in services['items']:
+                service_type = service['spec'].get('type')
+                if service_type != "ClusterIP" and service_type != "NodePort":
+                    continue
+
+                service_ip = service['spec'].get('clusterIP')
+                if not service_ip:
+                    continue
+
+                service_ports = service['spec'].get('ports')
+                if not service_ports:
+                    continue
+
+                external_ips = service['spec'].get('externalIPs')
+
+                for service_port in service_ports:
+                    if service_type == "NodePort":
+                        port = service_port.get('nodePort')
+                    else:
+                        port = service_port.get('port')
+
+                    if not port:
+                        continue
+
+                    protocol = service_port.get('protocol', 'TCP')
+
+                    if service_type == "NodePort":
+                        physical_gateway_ips = self._get_physical_gateway_ips()
+                        for gateway_ip in physical_gateway_ips:
+                            key = "%s:%s" % (gateway_ip, port)
+                            if protocol == "TCP":
+                                tcp_nodeport_services.append(key)
+                            else:
+                                udp_nodeport_services.append(key)
+                    elif external_ips:
+                        for external_ip in external_ips:
+                            key = "%s:%s" % (external_ip, port)
+                            if protocol == "TCP":
+                                tcp_nodeport_services.append(key)
+                            else:
+                                udp_nodeport_services.append(key)
+                    else:
+                        key = "%s:%s" % (service_ip, port)
+                        if protocol == "TCP":
+                            tcp_services.append(key)
+                        else:
+                            udp_services.append(key)
+
+        # For each of the OVN load-balancer, if the VIP that exists in
+        # the load balancer is not seen in current k8s services, we
+        # delete it.
+        load_balancers = {variables.K8S_CLUSTER_LB_TCP: tcp_services,
+                          variables.K8S_CLUSTER_LB_UDP: udp_services,
+                          variables.K8S_NS_LB_TCP: tcp_nodeport_services,
+                          variables.K8S_NS_LB_UDP: udp_nodeport_services}
+
+        for load_balancer in load_balancers:
+            k8s_services = load_balancers[load_balancer]
+
+            vips = self._get_load_balancer_vips(load_balancer)
+            if not vips:
+                continue
+
+            for vip in vips:
```

Comment:
This is just a question - and I don't remember if you have already answered it.

Here we are syncing by removing ovn load balancer which have no corresponding services.
Is there a need to resync ovn load balancers whose set of endpoints has changed while the watcher was down? Or will they be synchronised automatically when ADDED events for endpoints are processed?